Kubernixos in golang

johanot opened the pull request:
This PR encompass the Golang rewrite of kubernixos.
The Golang nix shell dev environment has been "borrowed" from morph.

Please refer to the updated README for a description of usage and workings of kubernixos.

Commit:
Added golang project

Commit:
Remove bash version of kubernixos

Commit:
Fix packaging

Commit:
Updated README.md to fit new version of kubernixos

Commit:
Added TravisCI config equal to that of morph

adamtulinius commented on README.md:
```diff
@@ -1,18 +1,74 @@
+This will mutate nothing.
+
+`kubernixos apply`: Like kubectl apply. This will apply kubernixos state to the cluster.
+Unwanted objects will not be pruned, unless "prune" is added to the command explicitly.
```
Is `kubernixos apply --prune` a thing?

johanot replied:
Yes.. But here be dragons. If you do that, you opt in to the prune mechanism of kubectl, which.. you probably don't want :P Do you think we should block this somehow?

adamtulinius commented on Gopkg.toml:
```diff
@@ -0,0 +1,30 @@
+# [prune]
+#   non-go = false
+#   go-tests = true
+#   unused-packages = true
```
Is there any reason to keep the example?

johanot replied:
That is... auto-generated by `dep init`.. I see no harm in keeping what a proven tool has put there.

adamtulinius commented on kubernixos.go:
```diff
@@ -0,0 +1,146 @@
+	passthroughArgs := parseCmdline(os.Args[1:])
+
+	deployFile, err := ioutil.TempFile("", "kubernixos")
+	if err != nil {
```
Superfluous `if err != nil`?

adamtulinius commented on shell.nix:
```diff
@@ -0,0 +1,63 @@
+      if [[ `${which} kubernixos 2>&1 >/dev/null` ]]; then
+        export PATH=$PATH:$(pwd)/result-bin/bin
+      fi
+      #source <(morph --completion-script-bash)
```
Hmm? :P

johanot replied:
c/p fail.. will remove it

adamtulinius commented on kubeclient/kubeconfig.go:
```diff
@@ -0,0 +1,37 @@
+func GetKubeConfig(server string) (*rest.Config, error) {
+	var kubeconfig *string
+	if home := homedir.HomeDir(); home != "" {
+		kubeconfig = flag.String("kubeconfig", filepath.Join(home, ".kube", "config"), "(optional) absolute path to the kubeconfig file")
```
note to self: how does this interact with flags passed to kubernixos?

johanot replied:
Nice catch! I think this should be changed since it relies on goflags, which is not used elsewhere. Probably another c/p / cargo culting mistake. Will change it.

adamtulinius commented on lib/kubernixos.nix:
```diff
@@ -2,36 +2,40 @@
+    inherit pkgs modules;
+  }).config.kubernixos;
+
+  kubernixos = with builtins; substring 0 32 (hashString "sha256" (toJSON cfg.manifests));
```
Why are we only using the first 32 characters of the hash?

johanot replied:
First couple of attemps failed because the kubernetes label value was too long. At that time, I was too lazy to look up the actual max length. Did it now: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set ... Looks like it's max 63 chars, while in this format full hash length is 65 chars.

Like said, I was too lazy to look up the max length. Instead just quickly decided that a "precision" of 32 chars would be more than sufficient.

johanot replied:
which.. actually is 33 chars now (from 0-32) ;-) I'm open for suggestions here? Should we go for absolute maximum (63 chars) or rather leave some room for encoding other information in there in the future?

Commit:
Addressing review comments:

- Removed reminiscence of "morph"
- Removed dependency on "flag" and honor KUBECONFIG env var only
- Removed superfluous if err ...
- Added some missing entries to dep lock file
- Disabled the use of --prune in kubernixos
- Set kubernixos hash length to 63 chars
- Changed misleading function name

johanot commented:
@adamtulinius Checkout https://github.com/DBCDK/kubernixos/pull/2/commits/cb9dbc194b67a47bb225a5f052e5e8abea21e5b3 .. This commit addresses some of your comments, and you should be able to resolve them right away.
